UserDataProcessingFindFailureReasonActivity/handler.ts

```diff
@@ -0,0 +1,133 @@
+
+// Activity failed because of invalid input
+export const ActivityResultInvalidInputFailure = t.interface({
+  kind: t.literal("INVALID_INPUT_FAILURE"),
+  reason: t.string
+});
+export type ActivityResultInvalidInputFailure = t.TypeOf<
+  typeof ActivityResultInvalidInputFailure
+>;
+
+// Activity failed because of invalid input
+export const ActivityResultNotFoundFailure = t.interface({
+  kind: t.literal("NOT_FOUND_FAILURE")
+});
+export type ActivityResultNotFoundFailure = t.TypeOf<
+  typeof ActivityResultNotFoundFailure
+>;
+
+// Activity failed because of invalid input
+export const ActivityResultUnhandledFailure = t.interface({
+  kind: t.literal("UNHANDLED"),
+  reason: t.string
+});
+export type ActivityResultUnhandledFailure = t.TypeOf<
+  typeof ActivityResultUnhandledFailure
+>;
+
+// Activity failed because of an error on a query
+export const ActivityResultQueryFailure = t.intersection([
+  t.interface({
+    kind: t.literal("QUERY_FAILURE"),
+    reason: t.string
+  }),
+  t.partial({ query: t.string })
+]);
+export type ActivityResultQueryFailure = t.TypeOf<
+  typeof ActivityResultQueryFailure
+>;
+
+export const ActivityResultFailure = t.taggedUnion("kind", [
+  ActivityResultQueryFailure,
+  ActivityResultInvalidInputFailure,
+  ActivityResultNotFoundFailure,
+  ActivityResultUnhandledFailure
+]);
+export type ActivityResultFailure = t.TypeOf<typeof ActivityResultFailure>;
+
+// Activity result
+export const ActivityResultSuccess = t.interface({
+  kind: t.literal("SUCCESS"),
+  value: NonEmptyString
+});
+export type ActivityResultSuccess = t.TypeOf<typeof ActivityResultSuccess>;
+
+export const ActivityResult = t.taggedUnion("kind", [
+  ActivityResultSuccess,
+  ActivityResultFailure
+]);
+
+export type ActivityResult = t.TypeOf<typeof ActivityResult>;
+
+export const getFindFailureReasonActivityHandler = async (
+  context: Context,
+  input: unknown
+): Promise<ActivityResult> => {
+  const client = df.getClient(context);
+
+  return fromEither(ActivityInput.decode(input))
+    .mapLeft(_ => {
+      return ActivityResultFailure.encode({
+        kind: "INVALID_INPUT_FAILURE",
+        reason: "Input not valid for this activity"
+      });
+    })
+    .chain(({ choice, fiscalCode }) => {
+      // create failed orchestrator id based on choice
+      const failedUserDataProcessingOrchestratorId =
+        choice == UserDataProcessingChoiceEnum.DELETE
+          ? makeDeleteOrchestratorId(fiscalCode)
+          : makeDownloadOrchestratorId(fiscalCode);
+
+      return tryCatch(
+        () => {
+          // get the status of the failed orchestrator
+          return client.getStatus(failedUserDataProcessingOrchestratorId);
+        },
+        () => {
+          return ActivityResultFailure.encode({
+            kind: "NOT_FOUND_FAILURE"
+          });
+        }
+      );
+    })
+    .fold(identity, orchestratorStatus => {
+      return fromNullable("No reason found")(orchestratorStatus.output)
```

Comment:
What kind of output do you expect?

Reply:
I typed that output and decoded it, but in the end I thought that I needed a "string" reason that actually was saved in that "output" property. I take the string and save it as it is in the new reason. Do you prefer that I decode it as an ActivityFailure and then save a custom assembled string like "kind: reason"? Does this gives us any added value? What if for some strage reason there is not an ActivityFailure in that property and we cannot decode the string? Isn't better to have that string saved directly to the reason property?